Refactored Substitution to its own type

vladciobanu opened the pull request:
Added a new type `Substitution level variable` and replaced `UnificationSubstitution` pretty much everywhere. I don't really like `Kore.Unification.Substitution.unsafeWrap` but I fear the alternative might be more `hs-boot` files, so I'm not sure which is worse.

Added signatures to `ImplicitSorts.hs` because `hasktags` was failing for me.

I'm open to suggestions on how to make this better, but I am also eager to help with more urgent issues. I know we can do a lot more with substitutions, but this might not be the time for it.


###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

vladciobanu commented:
Jenkins: test this please.

virgil-serbanuta commented on src/main/haskell/kore/src/Kore/Unification/Substitution.hs:
```diff
@@ -0,0 +1,110 @@
+       ( StepPattern )
+
+data Substitution level variable
+    = Substitution ![(variable level, StepPattern level variable)]
```
My main problem with this is that we are sometimes representing unnormalized substitutions as predicates, e.g. when we unify `concat(elem(x, y), z))` with a map. By default I would have expected to have a (un)normalized substitution-predicate, but this approach may also make sense. If that's the case, could you add some documentation (which you should have for all public stuff anyway) explaining, among others, the expected usage?

virgil-serbanuta commented on src/main/haskell/kore/src/Kore/Unification/SubstitutionNormalization.hs:
```diff
@@ -72,20 +73,29 @@ normalizeSubstitution
         (PredicateSubstitution level variable)
 normalizeSubstitution tools substitution =
-    ExceptT . sequence . fmap maybeToBottom $ topologicalSortConverted
+    -- if Substitution.isNormalized substitution
```
Do you need this comment?

vladciobanu commented on src/main/haskell/kore/src/Kore/Implicit/ImplicitSorts.hs:
```diff
@@ -12,9 +12,24 @@ Portability : POSIX
+import Kore.MetaML.AST
 
 -- TODO(virgil-serbanuta): Add tests for "defined but not used" symbols
+charMetaSort     :: Sort Meta
```
Nope, you're absolutely right. Seems I also forgot a couple of signatures! Fixed now.

vladciobanu commented on src/main/haskell/kore/src/Kore/Step/Simplification/Equals.hs:
```diff
@@ -224,74 +224,73 @@ makeEvaluate
-        , predicate = PredicateTrue
-        , substitution = []
+        , predicate = secondPred
+        , substitution = secondSubst
```
Seems view patterns did simplify the code a lot. Good catch!

vladciobanu commented on src/main/haskell/kore/src/Kore/Unification/Data.hs:
```diff
@@ -77,11 +55,6 @@ data UnificationProof level variable
     -- if we instantiate it using  ϕ1 = x, ϕ2 = y and ψ = (v = t2), we get
     -- |- x = t1 -> ((x = t2) -> (t1 = t2))
     -- by boolean manipulation, we can get
```
I somehow didn't copy all comments. Fixed now, thanks!

vladciobanu commented:
Jenkins: test this please.

virgil-serbanuta commented on src/main/haskell/kore/src/Kore/Step/BaseStep.hs:
```diff
@@ -466,10 +470,10 @@ stepWithAxiomForUnifier
                 -- remove them?
                 , substitution =
-                    mapSubstitutionVariables
+                    Substitution.mapSubstitutionVariables
```
Since you're using it qualified, you could call it 'mapVariables'.

virgil-serbanuta commented on src/main/haskell/kore/src/Kore/Unification/UnifierImpl.hs:
```diff
@@ -197,13 +200,14 @@ solveGroupedSubstitution tools substitutionSimplifier var patterns = do
   where
     termAndSubstitution s =
-        (var, ExpandedPattern.term s) : ExpandedPattern.substitution s
+        (var, ExpandedPattern.term s) :
```
I think that the style guide requires the operator to be on the next line. I may be wrong, but if you're not moving the operator you should probably indent the next line.

virgil-serbanuta commented on src/main/haskell/kore/src/Kore/Unification/UnifierImpl.hs:
```diff
@@ -249,9 +253,8 @@ normalizeSubstitutionDuplication tools substitutionSimplifier subst =
-                     ++ Predicated.substitution predSubst
-                    )
+                    $ (Substitution.wrap $ concat singletonSubstitutions)
+                     <> Predicated.substitution predSubst
```
Strange indentation level, I'd use 3 more spaces here (admittedly, it was somewhat strange in the original version)

virgil-serbanuta commented on src/main/haskell/kore/src/Kore/Unification/Substitution.hs:
```diff
@@ -0,0 +1,121 @@
+{-|
```
It would be nice to have a unit test for this file.

vladciobanu commented:
Thank you @virgil-serbanuta , I addressed all comments except the unit tests. I am eager to merge this because I already had to deal with a lot of merge conflicts. I added [this card](https://trello.com/c/x9Fv0yWv/361-write-unit-tests-for-the-new-substitution-module).

ttuegel commented on src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs:
```diff
@@ -471,15 +472,19 @@ matchNonVarToPattern tools substitutionSimplifier first second
+                Predicated
+                    { term = ()
+                    , predicate
+                    , substitution = Substitution.wrap rightSubst }
```
I think our style guide would demand that the closing brace go on the next line:
```suggestion
                    , substitution = Substitution.wrap rightSubst
                    }
```
